### PR TITLE
198 partner API - API key migration

### LINKED ISSUE #198: Partner API
Įgyvendintas individualių API raktų valdymas, leidžiantis institucijai pagal poreikį generuoti/pergeneruoti individualius  API raktus, perkelti dabar institucijų naudojami raktai.

---

Perkelti Parnerių API:

https://data.gov.lt/partner/api/1

API turi būti identiškai toks pat, kadangi jį naudoja partneriai, todėl po naujos versijos išleidimo integracijos turi veikti, kaip veikusios.

Iki šiol buvo išduodamas vienas API raktas visai organizacijai, tačiau reikėtų perdaryti, kad API raktas būtų išduodamas kiekvienam `Representative` ir gali būti keičiamas `Representative` formoje, pažymėjus checkbox „Suteikti API prieigą“. Jei prieiga suteikta, generuojamas raktas ir atvaizduojamas formoje, kaip read-only laukas. Jei checkbox „Suteikti API prieigą“ yra nepažymėtas, raktas pašalinamas.

Turėtu būti dar vienas checkbox „Pergeneruoti raktą“, kurį pažymėjus, pergeneruojamas raktas.

Padaryti migraciją, kad visiems organizacijos representative suteikiamas vienodas organizacijai suteiktas api raktas.

Per API pateikti duomenys validuojami su formomis.

Rekomenduoju šią užduotį pateikti ne vienu, o keliais atskirais pull request, pirmiausiai padarius autorizacijos raktų migraciją, po to įgyvendinus vieną veikiantį API endpointą, o po to savo nuožiūra norimo dydžio gabalais.

### PR DESCRIPTION
**Pull request by adp-atea**

### CODE REVIEW

**Comment by adp-atea** [comment]
As the task required, there will be couple pull requests. This is the first pull request with API key migration and option to generate API keys in Representative form.

File: vitrina/api/migrations/0003_apikey_representative.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            for rep in Representative.objects.filter(content_type=ct, object_id=key.organization.pk):
+                ApiKey.objects.create(
+                    version=key.version,
+                    api_key=key.api_key,
+                    enabled=key.enabled,
+                    expires=key.expires,
+                    representative=rep
+                )
+                rep.has_api_access = True
+                rep.save()
```
Comment (sirex) [inline vitrina/api/migrations/0003_apikey_representative.py:26]: Delete all ApiKey's where representative is not set. There will no longer be organization-wide keys.

File: vitrina/orgs/views.py [hunk 2]
```diff
@@ -180,6 +183,14 @@ def form_valid(self, form):
             )
             messages.info(self.request, _("Naudotojui išsiųstas laiškas dėl registracijos"))
         self.object.save()
+
+        if self.object.has_api_access:
+            ApiKey.objects.create(
+                api_key=str(uuid.uuid4()),
```
Comment (sirex) [inline vitrina/orgs/views.py]: Previously for secret keys uuid4 was used, but it is not really meant for this, So I suggest to use more secure [secrets.token_urlsafe()](https://docs.python.org/3/library/secrets.html#secrets.token_urlsafe) function.

```suggestion
                api_key=secrets.token_urlsafe(),
```

**Review by sirex — CHANGES_REQUESTED** [review]
Mostly looks good, just minor comments.

File: vitrina/api/migrations/0003_apikey_representative.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+def delete_non_representative_api_keys(apps, schema_editor):
+    ApiKey = apps.get_model('vitrina_api', 'ApiKey')
+
+    for key in ApiKey.objects.all():
+        if not key.representative:
+            key.delete()
```
Comment (sirex) [inline vitrina/api/migrations/0003_apikey_representative.py]:
```suggestion
    ApiKey.objects.filter(representative__isnull=True).delete()
```

Electricity is expensive these days, so please, do not burn CPU if this is not needed.

**Review by sirex — APPROVED** [review]

**Pull request merged**